File: final_projects/ROS1_to_ROS2_Migration_Project/project_ros1_noetic/src/turtlesim_project_py/scripts/turtle_controller.py
#!/usr/bin/env python3
import rospy
import math
from my_robot_msgs.msg import Coordinates2D
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose


class TurtleController:

    def __init__(self):
        self.pose_ = None
        self.target_ = None
        self.cmd_vel_publisher_ = rospy.Publisher(
            "/turtle1/cmd_vel", Twist, queue_size=10)
        self.pose_subscriber_ = rospy.Subscriber(
            "/turtle1/pose", Pose, self.callback_turtle_pose, queue_size=10)
        self.target_subscriber_ = rospy.Subscriber(
            "target_coordinates", Coordinates2D, self.callback_target)
        self.control_loop_timer_ = rospy.Timer(
            rospy.Duration(0.01), self.control_loop)
        rospy.loginfo("Turtle controller has been started.")

    def callback_turtle_pose(self, msg):
        self.pose_ = msg

    def callback_target(self, msg):
        if msg.x > 0.0 and msg.x < 11.0 and msg.y > 0.0 and msg.y < 11.0:
            rospy.loginfo("Received new valid target")
            self.target_ = msg

    def control_loop(self, event=None):
        if self.pose_ == None or self.target_ == None:
            return

        dist_x = self.target_.x - self.pose_.x
        dist_y = self.target_.y - self.pose_.y
        distance = math.sqrt(dist_x * dist_x + dist_y * dist_y)

        msg = Twist()

        if distance > 0.5:
            # position
            msg.linear.x = 2*distance

            # orientation
            goal_theta = math.atan2(dist_y, dist_x)
            diff = goal_theta - self.pose_.theta
            if diff > math.pi:
                diff -= 2*math.pi
            elif diff < -math.pi:
                diff += 2*math.pi

            msg.angular.z = 6*diff
        else:
            # target reached!
            msg.linear.x = 0.0
            msg.angular.z = 0.0
            self.target_ = None

        self.cmd_vel_publisher_.publish(msg)


if __name__ == '__main__':
    rospy.init_node('turtle_controller')
    controller = TurtleController()
    rospy.spin()

File: INTERNSHIP/final_projects/ROS1_to_ROS2_Migration_Project/project_ros1_noetic/src/turtlesim_project_cpp/src/turtle_controller.cpp
#include <ros/ros.h>
#include <my_robot_msgs/Coordinates2D.h>
#include <geometry_msgs/Twist.h>
#include <turtlesim/Pose.h>

class TurtleController
{

public:
    TurtleController(ros::NodeHandle *nh) : turtlesim_up_(false), target_up_(false)
    {
        cmd_vel_publisher_ = nh->advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10);
        pose_subscriber_ = nh->subscribe("/turtle1/pose", 10, &TurtleController::callbackTurtlePose, this);
        target_subscriber_ = nh->subscribe("target_coordinates", 10, &TurtleController::callbackTarget, this);
        control_loop_timer_ = nh->createTimer(ros::Duration(0.01), std::bind(&TurtleController::controlLoop, this));
        ROS_INFO("Turtle controller has been started.");
    }

private:
    turtlesim::Pose pose_;
    my_robot_msgs::Coordinates2D target_;
    bool turtlesim_up_;
    bool target_up_;

    ros::Publisher cmd_vel_publisher_;
    ros::Subscriber pose_subscriber_;
    ros::Subscriber target_subscriber_;
    ros::Timer control_loop_timer_;

    void callbackTurtlePose(const turtlesim::Pose &msg)
    {
        pose_ = msg;
        turtlesim_up_ = true;
    }

    void callbackTarget(const my_robot_msgs::Coordinates2D &msg)
    {
        if (msg.x > 0.0 && msg.x < 11.0 && msg.y > 0.0 && msg.y < 11.0)
        {
            ROS_INFO("Received new valid target");
            target_ = msg;
            target_up_ = true;
        }
    }

    void controlLoop()
    {
        if (!turtlesim_up_ || !target_up_)
        {
            return;
        }

        double dist_x = target_.x - pose_.x;
        double dist_y = target_.y - pose_.y;
        double distance = std::sqrt(dist_x * dist_x + dist_y * dist_y);

        geometry_msgs::Twist msg;

        if (distance > 0.5)
        {
            // position
            msg.linear.x = 2 * distance;

            // orientation
            double steering_angle = std::atan2(dist_y, dist_x);
            double angle_diff = steering_angle - pose_.theta;
            if (angle_diff > M_PI)
            {
                angle_diff -= 2 * M_PI;
            }
            else if (angle_diff < -M_PI)
            {
                angle_diff += 2 * M_PI;
            }
            msg.angular.z = 6 * angle_diff;
        }
        else
        {
            // target reached!
            msg.linear.x = 0.0;
            msg.angular.z = 0.0;
        }

        cmd_vel_publisher_.publish(msg);
    }
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "turtle_controller");
    ros::NodeHandle nh;
    TurtleController controller = TurtleController(&nh);
    ros::spin();
    return 0;
}